pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,494 @@
+package purger
+
+import (
+	"context"
+	"flag"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/weaveworks/common/user"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/purger/purgeplan"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+const millisecondPerDay = int64(time.Second)
+
+var separatorString = string([]byte{model.SeparatorByte})
+
+// DataPurgerConfig holds config for DataPurger
+type DataPurgerConfig struct {
+	NumWorkers int `yaml:"num_workers"`
+}
+
+// RegisterFlags registers CLI flags for DataPurgerConfig
+func (cfg *DataPurgerConfig) RegisterFlags(f *flag.FlagSet) {
+	f.IntVar(&cfg.NumWorkers, "num-workers", 2, "Number of workers executing delete plans in parallel")
+}
+
+type workerJob struct {
+	planNo          int
+	userID          string
+	deleteRequestID string
+}
+
+type workerJobExecutionStatus struct {
+	workerJob
+	err error
+}
+
+// DataPurger does the purging of data which is requested to be deleted
+type DataPurger struct {
```

Comment:
By instrumentation do you mean exposing metrics and traces? If yes then I haven't added them anywhere yet because I wanted to get the approach validated first.

Reply:
That's fine if you want to wait to implement them.